app/src/main/java/app/revanced/manager/ui/screen/PatcherScreen.kt

```diff
@@ -0,0 +1,433 @@
+
+    val patcherSucceeded by vm.patcherSucceeded.observeAsState(null)
+    val canInstall by remember { derivedStateOf { patcherSucceeded == true && (vm.installedPackageName != null || !vm.isInstalling) } }
+    var showInstallPicker by rememberSaveable { mutableStateOf(false) }
+
+    val steps by remember {
+        derivedStateOf {
+            vm.steps.groupBy { it.category }
+        }
+    }
+
+    if (showInstallPicker)
+        InstallPickerDialog(
+            onDismiss = { showInstallPicker = false },
+            onConfirm = vm::install
+        )
+
+    AppScaffold(
+        topBar = {
+            AppTopBar(
+                title = stringResource(R.string.patcher),
+                onBackClick = onBackClick
+            )
+        },
+        bottomBar = {
+            BottomAppBar(
+                actions = {
+                    IconButton(
+                        onClick = { exportApkLauncher.launch("${vm.packageName}.apk") },
+                        enabled = canInstall
+                    ) {
+                        Icon(Icons.Outlined.Save, stringResource(id = R.string.save_apk))
+                    }
+                    IconButton(
+                        onClick = { vm.exportLogs(context) },
+                        enabled = patcherSucceeded != null
+                    ) {
+                        Icon(Icons.Outlined.PostAdd, stringResource(id = R.string.save_logs))
+                    }
+                },
+                floatingActionButton = {
+                    AnimatedVisibility(visible = canInstall) {
+                        ExtendedFloatingActionButton(
+                            text = {
+                                Text(
+                                    stringResource(if (vm.installedPackageName == null) R.string.install_app else R.string.open_app)
+                                )
+                            },
+                            icon = {
+                                Icon(
+                                    Icons.Outlined.FileDownload,
+                                    stringResource(id = R.string.install_app)
+                                )
+                            },
+                            onClick = {
+                                if (vm.installedPackageName == null)
+                                    showInstallPicker = true
+                                else vm.open()
+                            }
+                        )
+                    }
+                }
+            )
+        }
+    ) { paddingValues ->
+        Column(
+            modifier = Modifier
+                .padding(paddingValues)
+                .fillMaxSize()
+        ) {
+            LinearProgressIndicator(
+                progress = {
+                    steps.flatMap { it.value }.let {
+                        it.count { step -> step.state == State.COMPLETED }.toFloat() / it.size.toFloat()
```

Comment:
Can `it.size` be zero?

Reply:
No because `steps` is assigned the output of `generateSteps()` so the list is never empty